### ISSUE TITLE
Prevent blank/empty lines from being added within a block when automatically removing an unnecessary else clause

### ISSUE THREAD

**Issue by PrincessRTFM** [eslint/eslint#12026]
**Tell us about your environment**

* **ESLint Version:** v6.1.0
* **Node Version:** v8.10.0
* **npm Version:** 3.5.2

**What parser (default, Babel-ESLint, etc.) are you using?** default

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

```yaml
---
env:
    browser: true
    es6: true
    jquery: true
rules:
    for-direction: error
    getter-return: error
    no-compare-neg-zero: error
    no-cond-assign:
        - error
        - except-parens
    no-constant-condition: error
    no-debugger: error
    no-dupe-args: error
    no-dupe-keys: error
    no-duplicate-case: error
    no-empty: error
    no-empty-character-class: error
    no-ex-assign: error
    no-extra-boolean-cast: warn
    no-extra-parens:
        - error
        - all
        - conditionalAssign: false
          nestedBinaryExpressions: false
          enforceForArrowConditionals: false
    no-extra-semi: warn
    no-func-assign: error
    no-inner-declarations: error
    no-invalid-regexp: error
    no-irregular-whitespace:
        - error
        - skipStrings: false
          skipComments: true
    no-misleading-character-class: error
    no-obj-calls: error
    no-prototype-builtins: warn
    no-regex-spaces: error
    no-sparse-arrays: error
    no-template-curly-in-string: error
    no-unexpected-multiline: error
    no-unreachable: error
    no-unsafe-finally: error
    no-unsafe-negation: error
    require-atomic-updates: error
    use-isnan: error
    valid-typeof: error
    accessor-pairs: error
    array-callback-return: error
    block-scoped-var: error
    class-methods-use-this: warn
    consistent-return:
        - warn
        - treatUndefinedAsUnspecified: true
    curly: error
    default-case: error
    dot-location:
        - error
        - property
    dot-notation:
        - error
        - allowPattern: "^_*[a-zA-Z]+(_[A-Za-z]*)*$"
    guard-for-in: error
    no-alert: warn
    no-caller: error
    no-case-declarations: error
    no-else-return: warn
    no-empty-function: warn
    "no-eq-null": error
    no-eval: error
    no-extend-native:
        - error
        - exceptions:
            - Array
            - String
    no-extra-bind: warn
    no-extra-label: warn
    no-fallthrough: error
    no-floating-decimal: error
    no-global-assign: error
    no-implicit-coercion:
        - error
        - string: false
          allow:
            - '!!'
    no-implied-eval: error
    no-invalid-this: error
    no-iterator: error
    no-lone-blocks: error
    no-loop-func: error
    no-multi-str: warn
    no-new: error
    no-new-func: error
    no-new-wrappers: error
    no-octal: warn
    no-octal-escape: error
    no-proto: error
    no-redeclare:
        - error
        - builtinGlobals: true
    no-return-assign: error
    no-return-await: error
    no-script-url: warn
    no-self-assign: error
    no-self-compare: error
    no-sequences: error
    no-throw-literal: warn
    no-unmodified-loop-condition: error
    no-unused-expressions:
        - error
        - allowTaggedTemplates: true
    no-unused-labels: error
    no-useless-call: warn
    no-useless-catch: error
    no-useless-concat: error
    no-useless-escape: warn
    no-useless-return: error
    no-with: error
    prefer-promise-reject-errors:
        - warn
        - allowEmptyReject: true
    radix: error
    require-await: error
    require-unicode-regexp: warn
    wrap-iife:
        - error
        - any
        - functionPrototypeMethods: true
    yoda: error
    no-delete-var: error
    no-label-var: error
    no-shadow: error
    no-shadow-restricted-names: error
    no-undef:
        - error
        - typeof: true
    no-undefined: warn
    no-unused-vars:
        - error
        - vars: local
          args: after-used
    no-use-before-define: error
    array-bracket-newline:
        - error
        - multiline: true
          minItems: 3
    array-element-newline:
        - error
        - multiline: true
          minItems: 3
    block-spacing: error
    brace-style:
        - error
        - stroustrup
        - allowSingleLine: true
    camelcase: error
    comma-dangle:
        - error
        - always-multiline
    comma-spacing: error
    comma-style: error
    computed-property-spacing: error
    consistent-this:
        - error
        - that
        - self
        - me
    eol-last: error
    func-call-spacing: error
    func-name-matching:
        - warn
        - always
        - considerPropertyDescriptor: true
    func-names:
        - error
        - as-needed
    function-paren-newline:
        - error
        - multiline-arguments
    implicit-arrow-linebreak: error
    indent:
        - error
        - tab
    key-spacing: error
    keyword-spacing: error
    linebreak-style: error
    lines-between-class-members:
        - error
        - never
    multiline-ternary:
        - error
        - always-multiline
    new-parens: error
    newline-per-chained-call:
        - error
        - ignoreChainWithDepth: 3
    no-array-constructor: error
    no-lonely-if: error
    no-mixed-operators: warn
    no-mixed-spaces-and-tabs:
        - error
        - smart-tabs
    no-multi-assign: error
    no-multiple-empty-lines: error
    no-negated-condition: warn
    no-new-object: error
    no-tabs:
        - error
        - allowIndentationTabs: true
    no-trailing-spaces: error
    no-unneeded-ternary: error
    no-whitespace-before-property: error
    object-curly-newline:
        - error
        - multiline: true
          minProperties: 2
          consistent: true
    object-curly-spacing: error
    object-property-newline:
        - error
        - allowAllPropertiesOnSameLine: false
    one-var:
        - error
        - never
    operator-assignment: error
    operator-linebreak: error
    padded-blocks:
        - error
        - never
    quote-props:
        - error
        - consistent
    semi: error
    semi-spacing:
        - error
        - before: false
          after: true
    semi-style: error
    space-before-blocks: error
    space-before-function-paren:
        - error
        - anonymous: never
          named: never
          asyncArrow: always
    space-in-parens: error
    space-infix-ops:
        - error
        - int32Hint: true
    space-unary-ops: error
    spaced-comment: error
    switch-colon-spacing: error
    template-tag-spacing: error
    wrap-regex: error
    arrow-body-style:
        - error
        - as-needed
    arrow-parens:
        - error
        - as-needed
    arrow-spacing: error
    constructor-super: error
    generator-star-spacing:
        - error
        - after
    no-class-assign: error
    no-confusing-arrow:
        - error
        - allowParens: true
    no-const-assign: error
    no-dupe-class-members: error
    no-new-symbol: error
    no-this-before-super: error
    no-useless-constructor: error
    no-useless-computed-key: error
    no-useless-rename: error
    no-var: error
    object-shorthand: error
    prefer-arrow-callback: error
    prefer-const: warn
    prefer-numeric-literals: warn
    prefer-rest-params: error
    prefer-spread: error
    prefer-template: error
    require-yield: error
    rest-spread-spacing: error
    symbol-description: error
    template-curly-spacing: error
    yield-star-spacing: error
```

</details>

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

```js
function checkUnicornReserves(resNumber, isPasture, currUps, ivoryNeeded) {
	let unicornsLeft = 0;
	if (!isPasture) {
		let tearsLeft = resNumber - game.resPool.get('tears').value;
		unicornsLeft = 2500 * Math.ceil(tearsLeft / getZiggurats());
	}
	else {
		unicornsLeft = resNumber;
	}
	unicornsLeft = unicornsLeft - game.resPool.get('unicorns').value;
	let ivoryLeft = ivoryNeeded - game.resPool.get('ivory').value;
	if (unicornsLeft > 0) {
		return `You need ${game.getDisplayValueExt(unicornsLeft)} more unicorns (approximately ${game.toDisplaySeconds(unicornsLeft / currUps)}) to build this.`;
	}
	if (ivoryLeft > 0){
		return "You have enough unicorns, but need more ivory to build this.";
	}
	else {
		return "You have enough resources to build this now.";
	}
}
```

```bash
eslint --fix AutoKittens.js
```

**What did you expect to happen?**
The last `else` block should have been removed and the contained `return` statement should have replaced it entirely, with indentation corrected. No blank lines should have been generated.

**What actually happened? Please include the actual, raw output from ESLint.**
The `else` block was replaced with a blank line, followed by the re-indented `return` statement, as shown here:

```js
function checkUnicornReserves(resNumber, isPasture, currUps, ivoryNeeded) {
	let unicornsLeft = 0;
	if (!isPasture) {
		const tearsLeft = resNumber - game.resPool.get('tears').value;
		unicornsLeft = 2500 * Math.ceil(tearsLeft / getZiggurats());
	}
	else {
		unicornsLeft = resNumber;
	}
	unicornsLeft -= game.resPool.get('unicorns').value;
	const ivoryLeft = ivoryNeeded - game.resPool.get('ivory').value;
	if (unicornsLeft > 0) {
		return `You need ${game.getDisplayValueExt(unicornsLeft)} more unicorns (approximately ${game.toDisplaySeconds(unicornsLeft / currUps)}) to build this.`;
	}
	if (ivoryLeft > 0) {
		return "You have enough unicorns, but need more ivory to build this.";
	}

	return "You have enough resources to build this now.";
}
```

**Are you willing to submit a pull request to fix this bug?**
I do not have the necessary Javascript skill to do so.

**Comment by platinumazure** [comment]
Hi @PrincessRTFM, thanks for the issue.

By design, ESLint rules do not know about each other, but we do sometimes write autofixes to be as minimal as possible without speculating about things like indentation. This is because users can use multiple rules in their config, and thanks to our multiple-pass autofix, other rules can also apply fixes.

In your case, you should be able to get what you want by enabling and configuring the `indent` and `padding-line-between-statements` rules. Please give that a try and let us know if that works for you.

**Comment by PrincessRTFM** [comment]
I do have `indent` enabled and configured, but I do not have `padding-line-between-statements` turned on, because I do like having blank line _outside_ of indented blocks. It looks like this may just not be avoidable. I opened the issue only because it was introducing an additional blank line where previously there hadn't been one.

**Comment by platinumazure** [comment]
I'm not opposed to considering changing the rule to avoid introducing a blank line, but I think we have the rule in its current configuration because most people like having a line break after an `if` statement (whether it has an `else` or not), and we assume that `padding-line-between-statements` can be used in exceptional cases.

**Comment by PrincessRTFM** [comment]
Understandable. I personally dislike any blank lines inside indented blocks, and I prefer to keep `else` immediately following the end of the `if` block, so perhaps I'm simply one of the minority on that.

However, if such a line break is desired, could they not write one in themselves? Eg,

```js
if (condition) {
	// code to execute
	return value;
}

else {
	// different code
}
```

In that case, directly replacing the entire `else` block with the following code (re-indented) would leave the obviously-desired blank line while not automatically introducing one for all such cases, including where a blank line is intentionally omitted. The current configuration doesn't leave the option of whether to include a blank line to the user; if the rule is enabled, a blank line seemingly _will_ be inserted, which then necessitates manually cleaning them out if they aren't desired.
